src/main/groovy/com/google/protobuf/gradle/ToolsLocator.groovy

```diff
@@ -0,0 +1,104 @@
+ * and/or other materials provided with the distribution.
+ *
+ * 3. Neither the name of the copyright holder nor the names of its contributors
+ * may be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package com.google.protobuf.gradle
+
+import org.gradle.api.GradleException
+import org.gradle.api.NamedDomainObjectContainer
+import org.gradle.api.Project
+import org.gradle.api.artifacts.Configuration
+import org.gradle.api.artifacts.Dependency
+
+/**
+ * Holds locations of all external executables, i.e., protoc and plugins.
+ */
+class ToolsLocator {
+
+  private final Project project
+  private final ExecutableLocator protoc
+  private final NamedDomainObjectContainer<ExecutableLocator> plugins
+
+  ToolsLocator(Project project) {
+    this.project = project
+    protoc = new ExecutableLocator('protoc')
+    plugins = project.container(ExecutableLocator)
+  }
+
+  /**
+   * For every ExecutableLocator that points to an artifact spec, resolves the
+   * spec, downloads the artifact, and point to the local path.
+   */
+  void resolve() {
+    resolve(protoc)
+    if (protoc.path == null) {
+      protoc.path = 'protoc'
+    }
+    plugins.each { plugin ->
+      resolve(plugin)
+      if (plugin.path == null) {
+        plugin.path = "protoc-gen-${plugin.name}"
+      }
+    }
+  }
+
+  void resolve(ExecutableLocator locator) {
+    if (locator.artifact != null) {
+      Configuration config = project.configurations.create(locator.name) {
+        visible = false
+        transitive = false
+        extendsFrom = []
+      }
+      def groupId, artifact, version
+      (groupId, artifact, version) = locator.artifact.split(":")
+      def notation = [group: groupId,
+                      name: artifact,
+                      version: version,
+                      classifier: project.osdetector.classifier,
+                      ext: 'exe']
+      project.logger.info("Resolving artifact: ${notation}")
+      Dependency dep = project.dependencies.add(config.name, notation)
+      Set<File> files = config.files(dep)
```

Comment:
Why would you need to catch the IllegalStateException? It seems that with the current code you can prove that it won't be thrown. If the artifact itself isn't found that will be an exception during files(). The file extension also seems guaranteed, since that it was specified. Basically, wrong number of files and wrong extension seem like AssertionErrors to me; errors that I may check, but that I would claim are impossible to trigger.